src/plugin-traverser.ts

```diff
@@ -0,0 +1,115 @@
+import * as fs from 'fs';
+import * as path from 'path';
+
+import { BasePlugin } from './abstract/base-plugin.plugin';
+import { PluginModuleOptions } from './interfaces/plugin-module-options.interface';
+import { PluginOptions } from './interfaces/plugin-options.interface';
+import { Provider } from '@nestjs/common';
+
+const node_modules = require('node_modules-path');
+
+export class PluginTraverser {
+  private _pluginType: string;
+  private _directories: Array<string> = [path.resolve(node_modules())];
+
+  constructor(pluginModuleOptions: PluginModuleOptions) {
+    this._pluginType = pluginModuleOptions.pluginType;
+    if (
+      pluginModuleOptions.directories &&
+      pluginModuleOptions.directories.length > 0
+    )
+      this._directories = pluginModuleOptions.directories;
+  }
+
+  public traverseDirectories(): Array<Provider<BasePlugin>> {
+    let modules = [];
+    this._directories.forEach((parentDirectory) => {
+      if (fs.existsSync(parentDirectory)) {
+        fs.readdirSync(parentDirectory, { withFileTypes: true })
+          .filter((dirent) => dirent.isDirectory())
+          .map((dirent) => dirent.name)
+          .filter((directoryName) =>
+            this.packageJsonExists(parentDirectory, directoryName),
+          )
+          .forEach((directoryName) => {
+            modules = [
+              ...modules,
+              ...this.processDirectory(parentDirectory, directoryName),
+            ];
+          });
+      }
+    });
+    return modules;
+  }
+
+  private processDirectory(
+    parentDirectory: string,
+    directoryName: string,
+  ): Array<BasePlugin> {
+    const packageJsonPath = this.createPackageJsonPath(
+      parentDirectory,
+      directoryName,
+    );
+    if (this.isPluginModule(this.parsePackageJson(packageJsonPath))) {
+      return this.importModule(
+        this.createModulePath(parentDirectory, directoryName),
+      );
+    }
+    return [];
+  }
+
+  private packageJsonExists(
+    parentDirectory: string,
+    directoryName: string,
+  ): boolean {
+    const packageJsonPath = this.createPackageJsonPath(
+      parentDirectory,
+      directoryName,
+    );
+    return fs.existsSync(packageJsonPath);
+  }
+
+  private createPackageJsonPath(
+    parentDirectory: string,
+    directoryName: string,
+  ): string {
+    return path.join(parentDirectory, directoryName, 'package.json');
+  }
+
+  private parsePackageJson(packageJsonPath: string): {
+    pluginModule: PluginOptions;
+  } {
+    return JSON.parse(fs.readFileSync(packageJsonPath, 'utf-8'));
+  }
+
+  private isPluginModule(packageJson: {
+    pluginModule: PluginOptions;
+  }): boolean {
+    return (
```

Comment:
Maybe we can apply some validation pattern here to warn user on logs that what's wrong with the plugin if we identified the module as plugin.